ingestion-beam/bin/generate-pioneer-keys

```diff
@@ -0,0 +1,21 @@
+#!/usr/bin/env python3
+
+import sys
+import json
+
+try:
+    from jwcrypto import jwk
+except ImportError as exc:
+    print(f"{exc.__class__.__name__}: run `pip install jwcrypto` and try again")
+    sys.exit(1)
+
+key = jwk.JWK.generate(kty="EC")
```

Comment:
Everything here looks in line with [the example key in the AET anon_id proposal](https://docs.google.com/document/d/1zH3eVVI_28Afg1JXe_McDrW4MTYuWhiJMQR6AQbli8I/edit#heading=h.cda7bwgmtch1) except that the generated keys need to have a `key_id`. This may not be necessary for pioneer, but I also don't think it would hurt to include here.

We can pass a `key_id` param to `generate` here to include a key_id. I think it would make sense to randomly generate that value. I'm not sure how mmiller generated the key_id for his example and I'm not sure what restrictions exist on format for that param, but it looks to be in the same format as the "x", "y", and "d" in the example.

So easiest thing here might be to generate a first key and steal its "x" value to insert into our real key.

```suggestion
key_id = json.loads(jwk.JWK.generate(kty="EC").export_public())['x']
key = jwk.JWK.generate(kty="EC", key_id=key_id)
```

Reply:
We've been using _a_ key id which is important to pioneer for routing and we should definitely investigate adding it here. But my assumption has been that the key id doesn't have any effect on encryption and is simply metadata (for routing, see [here](https://github.com/rhelmer/pioneer-v2-example/blob/master/manifest.json#L25) and [here](https://github.com/mozilla-services/cloudops-infra/commit/6c7d7ffe85686190645e728b7772bc00de76281a#diff-b412684ee04868a7b78e063ca9389cac), though they currently don't match). In pioneer the key id we've used is human-readable, and will usually map to the study name since we're using per-study keypairs.

I could be misunderstanding this and we've actually implemented our own key id mechanism separate from the existing one. 

At any rate, if this script spits out a key id in its parameters we can use that as the key id for pioneer too, though it makes some of the [logic I put together](https://github.com/mozilla-services/cloudops-infra/commit/6c7d7ffe85686190645e728b7772bc00de76281a#diff-3b42ed9e701b9a88d1891a516e6af5fe) moot for mapping semantic key ids.

Basically I'm fine with this key id being random, but we should redo our pioneer mechanisms to be like the aet-generated ones (i.e. _not_ having two separate key id mechanisms). I'm also fine with the id not being random if that turns out to make sense.

Reply:
The use of the [`kid` field is optional](https://tools.ietf.org/html/rfc7517#section-4.5) and doesn't follow any particular pattern. After reading the RFC for this particular field, I learned there is a JWK Set that is effectively the `KeyStore` object implemented in the decoder (minus logic for KMS). The `kid` field, if set and applicable, is used to select a key from the JWK Set.

In the case of Pioneer, the key doesn't need to be random, and it doesn't hurt to add the field to the key.

Reply:
It makes sense to me that AET's needs and Pioneer's needs are a bit different here. Pioneer has a concrete rationale for having a set of keys for different purposes, so meaningful names make sense in the Pioneer case. In the AET case, we really only need a single key so there's no meaningful name to give it, and an auto-generated name seems reasonable.

I do think we should embed the `kid` field in the keys so that the payloads clients send also have a `kid` field that we can use in case we end up with a need to disambiguate.